cabal-install/Distribution/Client/PackageHash.hs

```diff
@@ -211,18 +213,37 @@ instance Binary HashValue where
   put (HashValue digest) = put digest
   get = do
     digest <- get
-    unless (BS.length digest == 32) $ -- NB: valid for SHA256
-        fail "HashValue: bad digest"
+    -- Cannot do any sensible validation here. Although we use SHA256
+    -- for stuff we hash ourselves, we can also get hashes from TUF
+    -- and that can in principle use different hash functions in future.
     return (HashValue digest)
 
+-- | Hash some data. Currently uses SHA256.
+--
 hashValue :: LBS.ByteString -> HashValue
 hashValue = HashValue . SHA256.hashlazy
 
 showHashValue :: HashValue -> String
 showHashValue (HashValue digest) = BS.unpack (Base16.encode digest)
 
+-- | Hash the content of a file. Uses SHA256.
+--
 readFileHashValue :: FilePath -> IO HashValue
 readFileHashValue tarball =
     withBinaryFile tarball ReadMode $ \hnd ->
       evaluate . hashValue =<< LBS.hGetContents hnd
 
+-- | Convert a hash from TUF metadata into a 'PackageSourceHash'.
+--
+-- Note that TUF hashes don't neessarily have to be SHA256, since it can
+-- support new algorithms in future.
+--
+hashFromTUF :: Sec.Hash -> HashValue
+hashFromTUF (Sec.Hash hashstr) =
+    --TODO: [code cleanup] either we should get TUF to use raw bytestrings or
+    -- perhaps we should also just use a base16 string as the internal rep.
```

Comment:
It's not obvious to me that we want to enforce that it's always sha256. Our choices when getting hashes from TUF metadata are: die if there's no SHA256 hash (e.g. only SHA512 available in 3 years when experts tell us all to upgrade), use some best available hash (e.g. SHA512). It's nice if we use the same hash function for tarballs we encounter locally vs ones we get from repos, but it's not fundamental I think (it just causes rebuilds in some rare situations).

So for that reason I removed the length 32 check and allow us to get any suitable hash from hackage-security (which of course currently only supports SHA256 but it's explicitly designed to be flexible there).